File: server/EvalutionAlghoritm/SelectRanging.js
import { parentPort } from "worker_threads"

parentPort.on("message", (param) => {
  const res = SelectRanging(param);
  parentPort.postMessage(res);
})

function SelectRanging({ p_populations }) {
  let left = 0;
  let right = p_populations.length;
  let middle;
  let rand = Math.random();
  while (left <= right) {
    middle = Math.floor((left + right) / 2);
    if (rand >= p_populations[middle] && rand < p_populations[middle + 1]) {
      break;
    }
    else if (rand < p_populations[middle]) {
      right = middle - 1;
    }
    else if (rand >= p_populations[middle + 1]) {
      left = middle + 1;
    }
  }
  return middle;
}

File: server/EvalutionAlgoritmCpp/EvalutionAlgoritmCpp/EvulationAlgoritmCpp/SelectRanging.cpp
#include "SelectRanging.h"
#include "GetRndDouble.h"
int SelectRanging(vector<double> p_populations)
{
	int left = 0;
	int right = p_populations.size()-1;
	int middle;
	int rand = GetRndDouble();
	while(left<=right)
	{
		middle = (int)((left + right) / 2);
		if (rand >= p_populations[middle] && rand < p_populations[middle + 1])
			break;
		else if (rand < p_populations[middle])
			right = middle - 1;
		else if (rand >= p_populations[middle + 1])
			left = middle + 1;
	}
	return middle;

}